identity-iota/src/credential/presentation_validator.rs

```diff
@@ -0,0 +1,857 @@
+type PresentationValidationResult = std::result::Result<(), AccumulatedPresentationValidationError>;
+
+impl Default for PresentationValidator {
+  fn default() -> Self {
+    Self::new()
+  }
+}
+impl PresentationValidator {
+  /// Constructs a new [PresentationValidator].
+  pub fn new() -> Self {
+    Self {}
+  }
+
+  /// Validates the semantic structure of the [Presentation].
+  ///
+  /// # Terminology
+  /// This is a *validation unit*.
+  pub fn check_structure<U, V>(presentation: &Presentation<U, V>) -> Result<()> {
+    Self::check_structure_local_error(presentation).map_err(Error::UnsuccessfulValidationUnit)
+  }
+
+  fn check_structure_local_error<U, V>(presentation: &Presentation<U, V>) -> ValidationUnitResult {
+    presentation
+      .check_structure()
+      .map_err(ValidationError::PresentationStructure)
+  }
+
+  /// An iterator over the indices corresponding to the credentials that have the
+  /// `nonTransferable` property set, but the credential subject id does not correspond to URL of the presentation's
+  /// holder.
+  ///
+  /// The output of this iterator will always be a subset of
+  /// [Self::holder_not_subject_iter](Self::holder_not_subject_iter()).
+  pub fn non_transferable_violations<U, V>(presentation: &Presentation<U, V>) -> impl Iterator<Item = usize> + '_ {
+    Self::holder_not_subject_iter_internal(presentation)
+      .filter(|(_, credential)| credential.non_transferable.unwrap_or(false))
+      .map(|(position, _)| position)
+  }
+
+  /// Validates that the nonTransferable property is met.
+  ///
+  /// # Errors
+  /// Returns at the first credential requiring a nonTransferable property that is not met.
+  ///
+  /// If one needs to find *all* the nonTransferable violations of this presentation, then see
+  /// [Self::non_transferable_violations](Self::non_transferable_violations()).
+  ///
+  /// # Terminology
+  /// This is a *validation unit*
+  pub fn check_non_transferable<U, V>(presentation: &Presentation<U, V>) -> Result<()> {
+    Self::check_non_transferable_local_error(presentation).map_err(Error::UnsuccessfulValidationUnit)
+  }
+
+  fn check_non_transferable_local_error<U, V>(presentation: &Presentation<U, V>) -> ValidationUnitResult {
+    if let Some(position) = Self::non_transferable_violations(presentation).next() {
+      let err = ValidationError::InvalidHolderSubjectRelationship {
+        credential_position: position,
+      };
+      Err(err)
+    } else {
+      Ok(())
+    }
+  }
+
+  /// An iterator over indices corresponding to credentials where the credential subject id does not correspond to the
+  /// presentation's holder.
+  ///
+  /// The output of this iterator is always a superset of
+  /// [Self::non_transferable_violations]([Self::non_transferable_violations()]).
+  pub fn holder_not_subject_iter<U, V>(presentation: &Presentation<U, V>) -> impl Iterator<Item = usize> + '_ {
+    Self::holder_not_subject_iter_internal(presentation).map(|(position, _)| position)
+  }
```

Comment:
How does this translate to the Wasm bindings? Not sure if we want to expose this otherwise.

Reply:
Was thinking to return the `Array` from `js_sys` here. What do you think of that?

Reply:
This function is kind of the odd-one-out in terms of validation units, I would expect it to return an error itself.

E.g.
```rust
fn check_holder_not_subject(credential: &Credential, holder: &IotaDID/&Url) -> ValidationUnitResult;
```

Which is then called internally for each credential in the presentation.

Not sure if this is any better to be honest, just that it follows the current conventions more closely.

Reply:
I will just make this and `non_transferable_violations` private for now.